cpp/serve/engine_actions/new_request_prefill.cc

```diff
@@ -117,83 +120,130 @@ class NewRequestPrefillActionObj : public EngineActionObj {
+          cum_prefill_length += new_tokens.size();
         }
         RECORD_EVENT(trace_recorder_, rsentry->request->id, "finish embedding");
       }
 
       RECORD_EVENT(trace_recorder_, request_ids, "start prefill");
-      NDArray logits =
-          models_[model_id]->BatchPrefill(embeddings, request_internal_ids, prefill_lengths);
+      NDArray hidden_states = models_[model_id]->FuseEmbedHidden(
+          embeddings, hidden_states_for_input, /*batch_size*/ 1, /*seq_len*/ cum_prefill_length);
+      hidden_states = models_[model_id]->BatchPrefillToLastHidden(
+          hidden_states, request_internal_ids, prefill_lengths);
       RECORD_EVENT(trace_recorder_, request_ids, "finish prefill");
-      ICHECK_EQ(logits->ndim, 3);
-      ICHECK_EQ(logits->shape[0], 1);
-      ICHECK_EQ(logits->shape[1], num_rsentries);
+      ICHECK_EQ(hidden_states->ndim, 3);
+      ICHECK_EQ(hidden_states->shape[0], 1);
+      ICHECK_EQ(hidden_states->shape[1], cum_prefill_length);
 
       if (model_id == 0) {
         // We only need to sample for model 0 in prefill.
-        logits_for_sample = logits;
+        hidden_states_for_input = hidden_states;
       }
-    }
 
-    // - Update logits.
-    ICHECK(logits_for_sample.defined());
-    Array<GenerationConfig> generation_cfg;
-    Array<RequestModelState> mstates_for_logitproc;
-    generation_cfg.reserve(num_rsentries);
-    mstates_for_logitproc.reserve(num_rsentries);
-    for (int i = 0; i < num_rsentries; ++i) {
-      generation_cfg.push_back(prefill_inputs[i].rsentry->request->generation_cfg);
-      mstates_for_logitproc.push_back(prefill_inputs[i].rsentry->mstates[0]);
-    }
-    logits_for_sample = logits_for_sample.CreateView({num_rsentries, logits_for_sample->shape[2]},
-                                                     logits_for_sample->dtype);
-    logit_processor_->InplaceUpdateLogits(logits_for_sample, generation_cfg, mstates_for_logitproc,
-                                          request_ids);
-
-    // - Compute probability distributions.
-    NDArray probs_on_device =
-        logit_processor_->ComputeProbsFromLogits(logits_for_sample, generation_cfg, request_ids);
-
-    // - Sample tokens.
-    //   For rsentries which have children, sample
-    //   one token for each rstate that is depending.
-    //   Otherwise, sample a token for the current rstate.
-    std::vector<int> sample_indices;
-    std::vector<RequestStateEntry> rsentries_for_sample;
-    std::vector<RandomGenerator*> rngs;
-    sample_indices.reserve(num_rsentries);
-    rsentries_for_sample.reserve(num_rsentries);
-    rngs.reserve(num_rsentries);
-    request_ids.clear();
-    generation_cfg.clear();
-    for (int i = 0; i < num_rsentries; ++i) {
-      const RequestStateEntry& rsentry = prefill_inputs[i].rsentry;
-      // No sample for rsentries with remaining inputs.
-      if (!rsentry->mstates[0]->inputs.empty()) {
-        continue;
```

Comment:
Why are we removing these lines?

Reply:
They are moved into the loop nest. Because we need the base model to sample a token before prefilling the small models. The small models need the new token from base model.